classes/data-stores/ActionScheduler_DBStore.php

```diff
@@ -993,36 +994,90 @@ protected function claim_actions( $claim_id, $limit, ?DateTime $before_date = nu
 			$where  .= " AND group_id {$group_operator} ( $id_list )";
 		}
 
-		/**
-		 * Sets the order-by clause used in the action claim query.
-		 *
-		 * @since 3.4.0
-		 * @since 3.8.3 Made $claim_id and $hooks available.
-		 *
-		 * @param string $order_by_sql
-		 * @param string $claim_id Claim Id.
-		 * @param array  $hooks Hooks to filter for.
-		 */
-		$order    = apply_filters( 'action_scheduler_claim_actions_order_by', 'ORDER BY priority ASC, attempts ASC, scheduled_date_gmt ASC, action_id ASC', $claim_id, $hooks );
-		$params[] = $limit;
-
-		$sql           = $wpdb->prepare( "{$update} {$where} {$order} LIMIT %d", $params ); // phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared, WordPress.DB.PreparedSQLPlaceholders
-		$rows_affected = $wpdb->query( $sql ); // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared, WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching
-		if ( false === $rows_affected ) {
-			$error = empty( $wpdb->last_error )
-				? _x( 'unknown', 'database error', 'action-scheduler' )
-				: $wpdb->last_error;
+		$use_priorities = true;
+		$total_rows_affected = 0;
+		if ( $use_priorities ) {
+			$pending_priorities = $wpdb->get_col( $wpdb->prepare( "SELECT DISTINCT priority FROM {$wpdb->actionscheduler_actions} FORCE INDEX (claim_id_status_priority_scheduled_date_gmt) {$where} ORDER BY priority", $where_params ) );
+			if ( empty( $pending_priorities ) ) {
+				return 0;
+			}
 
-			throw new \RuntimeException(
-				sprintf(
+			foreach ( $pending_priorities as $pending_priority ) {
```

Comment:
This potentially loops over up to 255 priorities.  However, in the real world, it's somewhat unlikely that we would have a spread of a full 255 different levels of priorities being scheduled.  Even with the full potential range scheduled, testing has shown that it is still more efficient than the deadlocks caused by order by clauses.

If we find we need to reduce this range further, my first thoughts are to apply a filter when adding actions to limit the unique priorities that can be used.  I don't see a need for 255 options.